tests/core/pyspec/eth2spec/test/validator/test_validator_unittest.py

```diff
@@ -0,0 +1,296 @@
+
+
+@with_all_phases
+@spec_state_test
+def test_get_epoch_signature(spec, state):
+    block = spec.BeaconBlock()
+    privkey = privkeys[0]
+    pubkey = pubkeys[0]
+    signature = spec.get_epoch_signature(state, block, privkey)
+    domain = spec.get_domain(state, spec.DOMAIN_RANDAO, spec.compute_epoch_at_slot(block.slot))
+    signing_root = spec.compute_signing_root(spec.compute_epoch_at_slot(block.slot), domain)
+    assert bls.Verify(pubkey, signing_root, signature)
+
+
+@with_all_phases
+@spec_state_test
+def test_is_candidate_block(spec, state):
+    period_start = spec.SECONDS_PER_ETH1_BLOCK * spec.ETH1_FOLLOW_DISTANCE * 2 + 1000
+    run_is_candidate_block(
+        spec,
+        spec.Eth1Block(timestamp=period_start - spec.SECONDS_PER_ETH1_BLOCK * spec.ETH1_FOLLOW_DISTANCE),
+        period_start,
+        success=True,
+    )
+    run_is_candidate_block(
+        spec,
+        spec.Eth1Block(timestamp=period_start - spec.SECONDS_PER_ETH1_BLOCK * spec.ETH1_FOLLOW_DISTANCE + 1),
+        period_start,
+        success=False,
+    )
+    run_is_candidate_block(
+        spec,
+        spec.Eth1Block(timestamp=period_start - spec.SECONDS_PER_ETH1_BLOCK * spec.ETH1_FOLLOW_DISTANCE * 2),
+        period_start,
+        success=True,
+    )
+    run_is_candidate_block(
+        spec,
+        spec.Eth1Block(timestamp=period_start - spec.SECONDS_PER_ETH1_BLOCK * spec.ETH1_FOLLOW_DISTANCE * 2 - 1),
+        period_start,
+        success=False,
+    )
+
+
+@with_all_phases
+@spec_state_test
+def test_get_eth1_data_default_vote(spec, state):
+    min_new_period_epochs = get_min_new_period_epochs(spec)
+    for _ in range(min_new_period_epochs):
+        next_epoch(spec, state)
+
+    state.eth1_data_votes = ()
+    eth1_chain = []
+    eth1_data = spec.get_eth1_vote(state, eth1_chain)
+    assert eth1_data == state.eth1_data
+
+
+@with_all_phases
+@spec_state_test
+def test_get_eth1_data_consensus_vote(spec, state):
+    min_new_period_epochs = get_min_new_period_epochs(spec)
+    for _ in range(min_new_period_epochs):
+        next_epoch(spec, state)
+
+    period_start = spec.voting_period_start_time(state)
+    votes_length = spec.get_current_epoch(state) % spec.EPOCHS_PER_ETH1_VOTING_PERIOD
+    state.eth1_data_votes = ()
+    eth1_chain = []
+    eth1_data_votes = []
+    block = spec.Eth1Block(timestamp=period_start - spec.SECONDS_PER_ETH1_BLOCK * spec.ETH1_FOLLOW_DISTANCE)
+    for i in range(votes_length):
+        eth1_chain.append(block)
+        eth1_data_votes.append(spec.get_eth1_data(block))
+
+    state.eth1_data_votes = eth1_data_votes
+    eth1_data = spec.get_eth1_vote(state, eth1_chain)
+    assert eth1_data.block_hash == block.hash_tree_root()
+
+
+@with_all_phases
+@spec_state_test
+def test_compute_new_state_root(spec, state):
+    pre_state = state.copy()
+    post_state = state.copy()
+    block = build_empty_block(spec, state, state.slot + 1)
+    state_root = spec.compute_new_state_root(state, block)
```

Comment:
We should consider switching `compute_new_state_root` to use `state_transition(validate_result=False)` instead of call the two `process_*` functions directly

Reply:
but also fine to leave as is... yay stability :)

Reply:
Hmm just found something weird: we use `compute_` for the pure functions (I think?) But `compute_new_state_root` is not a pure function since it changes the given `state`. The previous sentence also says: "It is useful to be able to run a state transition function (working on a copy of the state)"
In that sense, for correctness, I think it's good to modify`compute_new_state_root` to

```python
def compute_new_state_root(state: BeaconState, block: BeaconBlock) -> Root:
    temp_state: BeaconState = state.copy()
    signed_block = SignedBeaconBlock(message=block)
    temp_state = state_transition(temp_state, signed_block, validate_result=False)
    return hash_tree_root(temp_state)
```

Reply:
looks good